File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2:
            return False
        open_br_stack: List[str]= []
        correspondent_open: Dict = {
            ")":"(",
            "}":"{",
            "]":"[",
            ">":"<",
        }
        close_brackets = correspondent_open.keys()
        open_brackets = correspondent_open.values()
        for br in s:
            if br in open_brackets:
                open_br_stack.append(br)
            elif br in close_brackets: 
                if open_br_stack and (open_br_stack[-1] == correspondent_open[br]):
                    open_br_stack.pop()#remove the most recent open bracket from stack
                else:
                    return False
        
        return True if not open_br_stack else False

File: 0020-valid-parentheses/0020-valid-parentheses.ts
function isValid(s: string): boolean {
  if (s.length % 2 ){
    return false;
  }
  let stack: string[] = []
  const open_bracket = ["(","[","{"];
  const close_bracket = {
    ")": "(",
    "]": "[",
    "}": "{",
  };
  for (let i=0; i<s.length; ++i){
    const c = s[i];
    if (open_bracket.some(br=>c===br)) {
      stack.push(c)
    } else if (close_bracket[c]=== stack[stack.length-1]) {
      stack.pop();
    } else {
      return false;
    }
  }
  return stack.length === 0;
};
